contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,254 @@
+# Explicit service links
+
+## Abstract
+
+Proposes an API for explicit links to services in the pod spec.
+
+## Motivation
+
+Currently, all pods are automatically injected with docker-link-style
+environment variables for each service in their namespaces.  Making service
+dependencies explicit in the PodSpec has a number of advantages:
+
+1.  Reduce accidental coupling
+2.  Reduce the number of IP tables rules that have to be created by the
+    kube-proxy
+3.  Identify relationships between pods and services, allowing user interfaces
+    to display additional information for end users.
+
+## Constraints and Assumptions
+
+1.  The old behavior must remain the default in the v1 API
+2.  It must be possible to migrate off the default to using explicit links
+
+## Use Cases
+
+1.  As a user, I want to be able to explicitly define which pods are injected
+    with information about a service so that I can ensure my pods receive
+    information only about services they actually need to use
+2.  As a user, I want to control how services are injected into pods and
+    containers
+
+### Explicit links between pods and services
+
+Currently in Kubernetes a pod is injected with docker- link style environment
+variables and simpler `{SERVICE_NAME}_SERVICE_HOST}` and
+`{SERVICE_NAME}_SERVICE_PORT`
+[environment variables](http://kubernetes.io/docs/user-guide/services/#environment-variables)
+for every service in its namespace.  This is disadvantageous for a number of
+reasons:
+
+1.  It is not possible to tell from the pod's spec which services it consumes
+2.  The proxy must generate rules on a node for every service in every namespace
+3.  There is no way to prevent accidental coupling
+
+Adding explicit links to services addresses all three of the above concerns,
+once the existing automatic injection behavior is turned off:
+
+1.  Explicit links will appear in the pod's spec
+2.  The proxy can be modified to generate rules only for the services which are
```

Comment:
Should we be concerned with kube-proxy update latency? Should there be some SLO for it?